core/src/main/scala/kafka/cluster/Replica.scala

```diff
@@ -36,25 +36,41 @@ class Replica(val brokerId: Int,
   // for local replica it is the log's end offset, for remote replicas its value is only updated by follower fetch
   @volatile private[this] var logEndOffsetMetadata: LogOffsetMetadata = LogOffsetMetadata.UnknownOffsetMetadata
 
+  // The log end offset value at the time the leader receives last FetchRequest from this follower.
+  // This is used to determine the last catch-up time of the follower
+  @volatile var lastFetchLeaderLogEndOffset: Long = 0L
+
+  // The time when the leader receives last FetchRequest from this follower
+  // This is used to determine the last catch-up time of the follower
+  @volatile var lastFetchTimeMs: Long = 0L
```

Comment:
It is used in `updateLogReadResult()`. When offset of fetch request >= lastFetchLeaderLogEndOffset, the lastCatchUpTime will be lastFetchTimeMs if it is current time.